libmamba/src/solv-cpp/pool.hpp

```diff
@@ -0,0 +1,214 @@
+
+        auto find_dependency(StringId name_id, RelationFlag flag, StringId version_id) const
+            -> std::optional<DependencyId>;
+        auto add_dependency(StringId name_id, RelationFlag flag, StringId version_id) -> DependencyId;
+        auto get_dependency_name(DependencyId id) const -> std::string_view;
+        auto get_dependency_version(DependencyId id) const -> std::string_view;
+        auto get_dependency_relation(DependencyId id) const -> std::string_view;
+        auto dependency_to_string(DependencyId id) const -> std::string;
+
+        void create_whatprovides();
+        template <typename UnaryFunc>
+        void for_each_whatprovides_id(DependencyId dep, UnaryFunc func) const;
+        template <typename UnaryFunc>
+        void for_each_whatprovides(DependencyId dep, UnaryFunc func) const;
+        template <typename UnaryFunc>
+        void for_each_whatprovides(DependencyId dep, UnaryFunc func);
+
+        auto select_solvables(const ObjQueue& job) const -> ObjQueue;
+
+        auto add_repo(std::string_view name) -> std::pair<RepoId, ObjRepoView>;
+        auto has_repo(RepoId id) const -> bool;
+        auto get_repo(RepoId id) -> std::optional<ObjRepoView>;
+        auto get_repo(RepoId id) const -> std::optional<ObjRepoViewConst>;
+        auto n_repos() const -> std::size_t;
+        auto remove_repo(RepoId id, bool reuse_ids) -> bool;
+        template <typename UnaryFunc>
+        void for_each_repo_id(UnaryFunc func) const;
+        template <typename UnaryFunc>
+        void for_each_repo(UnaryFunc func);
+        template <typename UnaryFunc>
+        void for_each_repo(UnaryFunc func) const;
+        auto installed_repo() const -> std::optional<ObjRepoViewConst>;
+        auto installed_repo() -> std::optional<ObjRepoView>;
+        void set_installed_repo(RepoId id);
+
+        auto get_solvable(SolvableId id) const -> std::optional<ObjSolvableViewConst>;
+        auto get_solvable(SolvableId id) -> std::optional<ObjSolvableView>;
+        template <typename UnaryFunc>
+        void for_each_solvable_id(UnaryFunc func) const;
+        template <typename UnaryFunc>
+        void for_each_solvable(UnaryFunc func) const;
+        template <typename UnaryFunc>
+        void for_each_solvable(UnaryFunc func);
+
+        template <typename Func>
+        void set_debug_callback(Func callback);
+
+    private:
+
+        struct PoolDeleter
+        {
+            void operator()(::Pool* ptr);
+        };
+
+        std::unique_ptr<void, void (*)(void*)> m_user_debug_callback;
+        // Must be deleted before the debug callback
+        std::unique_ptr<::Pool, ObjPool::PoolDeleter> m_pool = nullptr;
+    };
+}
+
+/*******************************
+ *  Implementation of ObjPool  *
+ *******************************/
+
+#include <solv/pool.h>
+
+namespace mamba::solv
+{
+
+    template <typename UnaryFunc>
+    void ObjPool::for_each_repo_id(UnaryFunc func) const
```

Comment:
Consider passing `UnaryFunc&&` here.

Reply:
Right, I never know which on is best since it done by value in the std lib.

Reply:
Yeah depends on a lot of things. Here I would have used forward reference as we directly invoke the function, we dont store it, we have no intention to copy it etc. But a copy is fine too, it's not a big deal.